man/rocq.1
.TH ROCQ 1
.
.SH NAME
rocq \- The Rocq Prover
.
.
.SH SYNOPSIS
.B rocq
[
.BI \-debug-shim
]
[
.BI \-v
.BI |
.BI \-\-version
]
[
subcommand
]
[
subcommand options
]
.
.SH DESCRIPTION
.
.B rocq
is the Rocq prover.
.
.SH SEE ALSO
.
.BR rocq (1),
.PP
.I
The Rocq Reference Manual.
.PP
The Rocq web site: http://coq.inria.fr


CHANGES
Changes from 8.8.2 to 8.9+beta1
===============================

Kernel

- Mutually defined records are now supported.

Notations

- New support for autonomous grammars of terms, called "custom
  entries" (see chapter "Syntax extensions" of the reference manual).

Tactics

- Added toplevel goal selector ! which expects a single focused goal.
  Use with Set Default Goal Selector to force focusing before tactics
  are called.

- The undocumented "nameless" forms `fix N`, `cofix` that were
  deprecated in 8.8 have been removed from LTAC's syntax; please use
  `fix ident N/cofix ident` to explicitly name the (co)fixpoint
  hypothesis to be introduced.

- Introduction tactics "intro"/"intros" on a goal which is an
  existential variable now force a refinement of the goal into a
  dependent product rather than failing.

- Support for fix/cofix added in Ltac "match" and "lazymatch".

- Ltac backtraces now include trace information about tactics
  called by OCaml-defined tactics.

- Option "Ltac Debug" now applies also to terms built using Ltac functions.

- Deprecated the Implicit Tactic family of commands.

- The default program obligation tactic uses a bounded proof search
  instead of an unbounded and potentially non-terminating one now
  (source of incompatibility).

- The `simple apply` tactic now respects the `Opaque` flag when called from
  Ltac (`auto` still does not respect it).

- Tactic `constr_eq` now adds universe constraints needed for the
  identity to the context (it used to ignore them). New tactic
  `constr_eq_strict` checks that the required constraints already hold
  without adding new ones. Preexisting tactic `constr_eq_nounivs` can
  still be used if you really want to ignore universe constraints.

- Tactics and tactic notations now understand the `deprecated` attribute.
- The `fourier` tactic has been removed. Please now use `lra` instead. You
  may need to add `Require Import Lra` to your developments. For compatibility,
  we now define `fourier` as a deprecated alias of `lra`.

- The `romega` tactics have been deprecated; please use `lia` instead.

Focusing

- Focusing bracket `{` now supports named goal selectors,
  e.g. `[x]: {` will focus on a goal (existential variable) named `x`.
  As usual, unfocus with `}` once the sub-goal is fully solved.

Standard Library

- Added `Ascii.eqb` and `String.eqb` and the `=?` notation for them,
  and proved some lemmas about them.  Note that this might cause
  incompatibilities if you have, e.g., string_scope and Z_scope both
  open with string_scope on top, and expect `=?` to refer to `Z.eqb`.
  Solution: wrap `_ =? _` in `(_ =? _)%Z` (or whichever scope you
  want).

- Added `Ndigits.N2Bv_sized`, and proved some lemmas about it.

- The scopes `int_scope` and `uint_scope` have been renamed to
  `dec_int_scope` and `dec_uint_scope`, to clash less with ssreflect
  and other packages.  They are still delimited by `%int` and `%uint`.

- Syntax notations for `string`, `ascii`, `Z`, `positive`, `N`, `R`,
  and `int31` are no longer available merely by `Require`ing the files
  that define the inductives.  You must `Import` `Coq.Strings.String`,
  `Coq.Strings.Ascii`, `Coq.ZArith.BinIntDef`, `Coq.PArith.BinPosDef`,
  `Coq.NArith.BinNatDef`, `Coq.Reals.Rdefinitions`, and
  `Coq.Numbers.Cyclic.Int31.Int31`, respectively, to be able to use
  these notations.  Note that passing `-compat 8.8` or issuing
  `Require Import Coq.Compat.Coq88` will make these notations
  available.  Users wishing to port their developments automatically
  may download `fix.py` from
  <https://gist.github.com/JasonGross/5d4558edf8f5c2c548a3d96c17820169>
  and run a command like `while true; do make -Okj 2>&1 |
  /path/to/fix.py; done` and get a cup of coffee.  (This command must
  be manually interrupted once the build finishes all the way though.
  Note also that this method is not fail-proof; you may have to adjust
  some scopes if you were relying on string notations not being
  available even when `string_scope` was open.)

- Numeral syntax for `nat` is no longer available without loading the
  entire prelude (`Require Import Coq.Init.Prelude`).  This only
  impacts users running Coq without the init library (`-nois` or
  `-noinit`) and also issuing `Require Import Coq.Init.Datatypes`.

Tools

- Coq_makefile lets one override or extend the following variables from
  the command line: COQFLAGS, COQCHKFLAGS, COQDOCFLAGS.
  COQFLAGS is now entirely separate from COQLIBS, so in custom Makefiles
  $(COQFLAGS) should be replaced by $(COQFLAGS) $(COQLIBS).

- Removed the gallina utility (extracts specification from Coq vernacular files).
  If you would like to maintain this tool externally, please contact us.

- Removed the Emacs modes distributed with Coq. You are advised to
  use Proof-General <https://proofgeneral.github.io/> (and optionally
  Company-Coq <https://github.com/cpitclaudel/company-coq>) instead.
  If your use case is not covered by these alternative Emacs modes,
  please open an issue. We can help set up external maintenance as part
  of Proof-General, or independently as part of coq-community.


Vernacular Commands

- Removed deprecated commands Arguments Scope and Implicit Arguments
  (not the option). Use the Arguments command instead.
- Nested proofs may be enabled through the option `Nested Proofs Allowed`.
  By default, they are disabled and produce an error. The deprecation
  warning which used to occur when using nested proofs has been removed.
- Added option Uniform Inductive Parameters which abstracts over parameters
  before typechecking constructors, allowing to write for example
  `Inductive list (A : Type) := nil : list | cons : A -> list -> list.`
- New Set Hint Variables/Constants Opaque/Transparent commands for setting
  globally the opacity flag of variables and constants in hint databases,
  overwritting the opacity set of the hint database.
- Added generic syntax for “attributes”, as in:
  `#[local] Lemma foo : bar.`

topbin/rocq.ml
(************************************************************************)
(*         *      The Rocq Prover / The Rocq Development Team           *)
(*  v      *         Copyright INRIA, CNRS and contributors             *)
(* <O___,, * (see version control and CREDITS file for authors & dates) *)
(*   \VV/  **************************************************************)
(*    //   *    This file is distributed under the terms of the         *)
(*         *     GNU Lesser General Public License Version 2.1          *)
(*         *     (see LICENSE file for the text of the license)         *)
(************************************************************************)

let fatal_error fmt = Printf.kfprintf (fun _ -> exit 1) stderr (fmt^^"%!")

let with_worker_gen opts basename args =
  Rocqshim.init opts args;
  let prog = Rocqshim.get_worker_path { package = "rocq-runtime"; basename } in
  let () = if opts.debug_shim then Printf.eprintf "Using %s\n%!" prog in
  let argv = Array.of_list (prog :: args) in
  Rocqshim.exec_or_create_process prog argv

let with_worker opts kind args =
  with_worker_gen opts "rocqworker" (("--kind="^kind) :: args)

let with_sibling_exe opts prog args =
  let prog = System.get_toplevel_path prog in
  let () = if opts.Rocqshim.debug_shim then Printf.eprintf "Using %s\n%!" prog in
  let argv = Array.of_list (prog :: args) in
  Rocqshim.exec_or_create_process prog argv

type subcommand =
  | Compile
  | Repl
  | ReplWithDrop
  | NativePrecompile
  | Check
  | Votour
  | PpMlg
  | Dep
  | Doc
  | Wc
  | Workmgr
  | Tex
  | Makefile
  | Timelog2Html

let subcommands = [
  ("compile", "Compile a Rocq source file", Compile);
  ("c", "Alias for compile", Compile);
  ("repl", "Interactive read-eval-print loop", Repl);
  ("top", "Alias for repl", Repl);
  ("repl-with-drop", "repl with Drop command", ReplWithDrop);
  ("top-with-drop", "Alias for repl-with-drop", ReplWithDrop);
  ("native-precompile", "Preprocess a compiled Rocq file for use by native_compute", NativePrecompile);
  ("check", "Check a compiled Rocq file (alias for command rocqchk)", Check);
  ("votour", "Low level debugging of compiled files", Votour);
  ("preprocess-mlg", "Preprocess Rocq grammar files (.mlg) to produce OCaml sources", PpMlg);
  ("pp-mlg", "Alias for preprocess-mlg", PpMlg);
  ("dep", "Print dependencies for compiling Rocq files", Dep);
  ("doc", "Generate documentation from a Rocq source file", Doc);
  ("wc", "Count lines of code in a Rocq source file", Wc);
  ("workmgr", "Control the number of parallel workers used by Rocq", Workmgr);
  ("tex", "Process Rocq code in a Latex document", Tex);
  ("makefile", "Generate a Makefile to compile a Rocq project", Makefile);
  ("timelog2html", "Combine timing information and a Rocq source file", Timelog2Html);
]

let print_usage fmt () =
  let longest_subcmd =
    List.fold_left (fun acc (cmd,_,_) -> max acc (String.length cmd)) 0 subcommands
  in
  let subcommands =
    List.map (fun (cmd,doc,_) ->
        let separator = String.make (3 + longest_subcmd - String.length cmd) ' ' in
        "    "^cmd^separator^doc)
      subcommands
  in
  Printf.fprintf fmt "Usage: rocq [-debug-shim] {-v|--version|--print-version|--help|SUBCOMMAND} [ARGUMENTS...]\n\
\n\
\n  -v, --version: print human readable version info\
\n  --print-version: print machine readable version info\
Supported subcommands:\n\
%s\n\
\n\
Use \"rocq subcommand --help\" to get more information about a given subcommand.\n" (String.concat "\n" subcommands)

let error_usage () =
  fatal_error "%a" print_usage ()

let run_subcommand opts args = function
  | Compile -> with_worker opts "compile" args
  | Repl -> with_worker opts "repl" args
  | ReplWithDrop -> with_worker_gen opts "rocqworker_with_drop" args
  | NativePrecompile -> with_worker_gen opts "rocqnative" args
  | Check -> with_sibling_exe opts "rocqchk" args
  | Votour -> with_sibling_exe opts "votour" args
  | PpMlg -> Coqpp_main.main args
  | Dep -> Coqdeplib.Rocqdep_main.main args
  | Doc -> Coqdoclib.Rocqdoc_main.main ~prog:(Sys.argv.(0) ^ " doc") args
  | Wc -> Rocqwc.main args
  | Workmgr -> Rocqworkmgr.main ~prog:(Sys.argv.(0) ^ " workmgr") args
  | Tex -> Rocqtex.main ~prog:(Sys.argv.(0) ^ " tex") args
  | Makefile -> Rocqmakefile.main ~prog:[Sys.argv.(0); "makefile"] args
  | Timelog2Html -> Benchlib.Timelog2html.main ~prog:(Sys.argv.(0) ^ " timelog2html") args

let () =
  if Array.length Sys.argv < 2 then error_usage ();

  let args = List.tl (Array.to_list Sys.argv) in
  let opts, args = Rocqshim.parse_opts args in
  match args with
  (* help prints *)
  | "-v" :: _ | "--version" :: _ -> Boot.Usage.version ()
  | ("-print-version"|"--print-version") :: _ -> Boot.Usage.machine_readable_version ()
  | ("-h" | "-H" | "-help" | "--help") :: _ ->
    Printf.printf "%a%!" print_usage ();
    exit 0

  | cmd :: args ->
    begin match List.find_opt (fun (cmd',_,_) -> String.equal cmd cmd') subcommands with
    | None -> fatal_error "Unknown subcommand %s\n%a%!" cmd print_usage ()
    | Some (_,_,cmd) -> run_subcommand opts args cmd
    end

  | [] -> error_usage ()
